keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+will give no guarantees about the kubeadm backend until it is
+announced as stable. README files in the kubeadm repository
+should denote that.
+
+It is out of scope for this proposal to try to solve the problem.
+
+#### The post-submit job for tagging/branching k/kubeadm fails
+
+In cases the post-submit job fails, maintainers should be able
+to apply the required changes manually or ask Prow maintainers
+to re-trigger the job. Sufficient number of kubeadm maintainers
+should have access and know how to perform these actions.
+
+Ideally the post-submit job should be setup to send an email
+notification on every failure at the SIG Cluster Lifecycle
+mailing list with a custom `testgrid` suffix:
+`kubernetes-sig-cluster-lifecycle+testgrid@googlegroups.com`
+This suffix is already used by kubeadm periodic jobs.
+
+Close to a release, coordination with SIG Release will be required,
+to make sure that the exact, desired SHA of kubeadm is released.
+
+#### The kubeadm go.mod includes outdated components
+
+Once kubeadm is decoupled from k/k it will no longer benefit
+from all the updates and security fixes that its components under
+"staging" see. The imports in the kubeadm go.mod will need
+to be updated on a regular basis to mitigate CVEs and critical bugs.
+
+Tools such as https://docs.renovatebot.com/golang/ were
+proposed. Automatic bulk updates can increase the noise in the
+kubeadm repository but always make sure kubeadm depends on the
+latest dependencies. Manual updates can get sloppy and miss critical
+commits.
+
+The current proposal is to rely on manual updates and make sure
+kubeadm is always up to date before a final release, but also invest
+time into automating this process.
+
+Coordination with the security team and relevant SIGs might be
+required.
+
+#### The kubeadm module does not use SEMVER
+
+Users importing the kubeadm module will find that the k8s versioning
+is actually not SEMVER, because breaking changes happen on MINOR
+changes instead of MAJOR ones.
+
+A proposed solution here is to dual-tag commits with a version that
+is SEMVER friendly - e.g.:
+- `v1.17.0` -> `v0.17.0`
+
+Dual-tagging with `v17.0.0` is not an option, because Go modules
+will require the `v17` package.
+
+But the problem remains with `v0.17.0`, because once k8s becomes
+`v2.0.0`, the dual-tags (SEMVER) will overlap with the non-SEMVER tags.
+
+There are no good solutions for this problem and users might have to
+use SHAs to import the kubeadm module.
+
+See the related discussions at:
+https://github.com/kubernetes/kubernetes/issues/72638
+https://github.com/kubernetes/kubernetes/issues/84372
+
+## Design Details
+
+### Enhancements proposals
+
+The current plan is for kubeadm to continue following the k/enhancements
+and KEP process, given the project will be part of the k8s release.
+There will be no extra burden on the release team as they are already
+following PRs and issues in repositories other than k/k.
```

Comment:
I suspect we do actually incur some need for synchronization here.

Something we've discovered recently is that the k/release tooling requires there be a new commit between a prior release and the current to be tagged and built release.  This naturally happens today for anything that is part of k/k.

Consider the case of no changes in k/kubeadm:  Today when we release tag k/k on master and release branches as a part of the release process there are (almost always) new commits on k/k relative to prior releases.  We've added something and we want to release that.  Looking at the history of release branch patch support though, there are regularly times when kubeadm has not changed.  When we do a patch release of k/k would we also rebuild and republish the non-changed kubeadm branch in order to have a release of kubeadm-X.YY.Z to match the k/k X.YY.Z release? That is the proposal and causes issues with tooling which is designed to not build when there's no change.  This could be addressed simply by adding the proposed KUBE_VERSION file via a new commit.  But when do we add that marker and why?  It can be early or late, manual or automatic...

Consider more broadly the case of changes on k/kubeadm:  When k/k is tagged at X.YY.(Z+1) we are always tagging a specific commit on the release branches.  The patch release team knows those commitids, because the release team has managed the patch content flowing into the release branches with the cherry-pick-approved process.  There is an explicit cherry-pick deadline and an implicit PR freeze (sometimes with patches merged deliberately still after the deadline) before the actual release cut.  This yields a relatively informal but pragmatic quality gate and control, while also allowing for things that need to merge as set and get some pre-release testing on the branch (ie: the branch is in a not-ready-to-release state at times) before release.  When the patch release team has determined a branch is "ready", then the (Z+1) release moves forward.    For an external repo, I see a few paths to achieve something similar, eg:

 * external branches are kept "always ready": this isn't always possible practically, in a non-race-free way.  If we miss, will we need to make a next k/k release in order to make the next k/kubeadm release including a missed commit?  Does splitting the monolith come to mean k/k runs at the cadence of the split components?
 * external branch manager communicates with and approves release team's intent to release, with
   blocking ability
 * external branch manager leaves some moving tag hint as to best preferred commit for next release: the idea of a `pre-<version>` tag is mentioned above, but is described as something that automation applies as a part of the release process, not an input designator for the release process.  If automation can apply that tag, it implies we're in an "always release ready" state on the branch, difficult to always achieve as already described.
 * external branch manager runs their branch with k/k-independent versioning and cadence, PRs update to a branched (pre k/k release-X.YY) yaml file in k/release indicating which version of their component is preferred in conjunction with that branch

We're definitely working to burn down the existing tooling and replace it with a simpler solution, with re-usable, testable components.  I don't offer its existing shortcomings as excuse or with the intent to limit an implementation.  Rather I call them out as now is the time for considering how new release tooling can mesh with managing external repos by design, since we're still in design and only earliest of implementation on point tools.  We have the opportunity to enable a better workflow than what we have today, or somewhat unnaturally tacking k/kubeadm support onto the side.  We shouldn't constrain this KEP by pre-conceived notions of the release tooling, since we have an opportunity right now to dramatically influence the shape of that tooling as we rebuild it from the ground up.

There is a basic problem we need to address in the shared solution: How does the release team / patch release team know what k/kubeadm code to release in conjunction with each and every k/k vX.YY.Z release?

Reply:
> Something we've discovered recently is that the k/release tooling requires there be a new commit between a prior release and the current to be tagged and built release. This naturally happens today for anything that is part of k/k.

yes i saw the discussions. this can be avoided if support for multiple tags per commit are handled well. i've seen this in other projects.

> Consider the case of no changes in k/kubeadm: Today when we release tag k/k on master and release branches as a part of the release process there are (almost always) new commits on k/k relative to prior releases. We've added something and we want to release that. Looking at the history of release branch patch support though, there are regularly times when kubeadm has not changed. When we do a patch release of k/k would we also rebuild and republish the non-changed kubeadm branch in order to have a release of kubeadm-X.YY.Z to match the k/k X.YY.Z release? That is the proposal and causes issues with tooling which is designed to not build when there's no change. This could be addressed simply by adding the proposed KUBE_VERSION file via a new commit. But when do we add that marker and why? It can be early or late, manual or automatic...

i think this is something that should be handled with multiple tags per-commit even if the contents do not change. the option to not iterate the version is worse because it will create a version miss match in a tarball, some comments having an older version:
- kubelet 1.17.5
- kube-apiserver 1.17.4
- kubeadm 1.17.1

i don't think the users and their infrastructure tools would enjoy handing that.

> There is an explicit cherry-pick deadline and an implicit PR freeze (sometimes with patches merged deliberately still after the deadline) before the actual release cut. This yields a relatively informal but pragmatic quality gate and control, while also allowing for things that need to merge as set and get some pre-release testing on the branch (ie: the branch is in a not-ready-to-release state at times) before release. 

the same quality gates will apply to k/kubeadm development.

> external branches are kept "always ready": this isn't always possible practically, in a non-race-free way. If we miss, will we need to make a next k/k release in order to make the next k/kubeadm release including a missed commit? Does splitting the monolith come to mean k/k runs at the cadence of the split components?

missing component will mean that the users need to build from source or wait for the next PATCH. the k8s train should not wait.

> external branch manager leaves some moving tag hint as to best preferred commit for next release: the idea of a pre-<version> tag is mentioned above, but is described as something that automation applies as a part of the release process, not an input designator for the release process. If automation can apply that tag, it implies we're in an "always release ready" state on the branch, difficult to always achieve as already described.

kubeadm as a project is low volume and stable enough to not require manual interaction.
also as mentioned earlier monitoring 4 branches and tagging them manually is not something that the kubeadm maintainers would like to do.

> We're definitely working to burn down the existing tooling and replace it with a simpler solution, with re-usable, testable components. I don't offer its existing shortcomings as excuse or with the intent to limit an implementation. Rather I call them out as now is the time for considering how new release tooling can mesh with managing external repos by design, since we're still in design and only earliest of implementation on point tools. We have the opportunity to enable a better workflow than what we have today, or somewhat unnaturally tacking k/kubeadm support onto the side. We shouldn't constrain this KEP by pre-conceived notions of the release tooling, since we have an opportunity right now to dramatically influence the shape of that tooling as we rebuild it from the ground up.

if k/release tooling is the excuse to not manage external repositories in the short term, then perhaps we should consider kubeadm to stop releasing as part of the k8s tarballs, packages and CI and simply build their own artifacts.

cri-tools are doing something like that except that k/kubeadm can do better.

> There is a basic problem we need to address in the shared solution: How does the release team / patch release team know what k/kubeadm code to release in conjunction with each and every k/k vX.YY.Z release?

currently, how do they know when k/k/cmd/kubeadm is ready for release?
we can continue having the same communication channels - slack and test-grid.
other than those i don't think they have any other way of knowing if kubeadm is in good shape.

Reply:
> currently, how do they know when k/k/cmd/kubeadm is ready for release?
> we can continue having the same communication channels - slack and test-grid.
> other than those i don't think they have any other way of knowing if kubeadm is in good shape.

For patch release it is implicit in k/k's cherry pick deadline having passed and nobody having raised a flag that it's not release ready.

For .0 release it is through discussion between the members of the release team, tracking testgrid, and additional conversation.

Reply:
> if k/release tooling is the excuse to not manage external repositories in the short term

As I explicitly said, let's not let this be an excuse.  What is the system we want which manages things well?  Let's build that!

Reply:
> As I explicitly said, let's not let this be an excuse. What is the system we want which manages things well? Let's build that!

yet, i think that the current state of k/release is concerning to the extend where we might not want to introduce more logic in there. logic such as conditionally building kubeadm from another repository based on a version TAG.

this was my biggest concern before starting to write the first words of this KEP.

Reply:
As with the the other threads, we can move ahead and experimentally run some on k/kubeadm release and see how things work.  I do believe things will probably work ok for kubeadm given how the project operates relative to the k/k release cycle, but also that this not-explicitly-sync'd workflow wont form the basis of a repeatable pattern for other components to pop out of k/k to their own repo and get the right content into "the release".